### PR TITLE
Progressive calls

### PR DESCRIPTION
**Pull request by KSDaemon**
This PR introduces Progressive calls WAMP Feature

### What kind of change does this PR introduce?
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:
- [x] My code follows the code style of this project.
- [ ] I have added tests to cover my changes.
- [ ] Overall test coverage is not decreased.
- [ ] All new and existing tests passed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by KSDaemon** [comment]
@bboozzoo Can you have a look? :)

File: client/client.go
Comment (zyga) [inline client/client.go:850]: What happens if the context gets cancelled? Will this naturally terminate because `progChan` gets closed?

File: client/client.go
Comment (zyga) [inline client/client.go]: Does this return when `ctx` gets cancelled?

File: client/client.go
Comment (zyga) [inline client/client.go]: Returns seems redundant here

File: client/client.go
Comment (zyga) [inline client/client.go]: Return seems redundant here

File: client/client.go
Comment (zyga) [inline client/client.go]: Return seems redundant here

File: client/client.go
Comment (zyga) [inline client/client.go]: Who is sending to `progChan`? Is that only done inside `waitForReplyWithCancel`?

File: client/client.go
Comment (zyga) [inline client/client.go]: You could `defer func() { if err != nil { finalrescb(nil, err) } }()` earlier, just after the call to `waitForReplyWithCancel` to ensure that the callback is called, this would also simplify the logic below.

File: client/client.go
Comment (zyga) [inline client/client.go]: Nitpick, combine this with error checking from the lines below.

File: client/client.go
Comment (zyga) [inline client/client.go]: I think you should just return `c.sess.Send(message)`, assuming return signatures match.

File: router/dealer.go
Comment (zyga) [inline router/dealer.go]: Nitpick, use `fmt.Sprintf` and either `%v` or `%q` as appropriate.

**Review by zyga — COMMENTED** [review]
I'm not very familiar with wamp so I asked a bunch of questions around concurrency and safety. I may send more questions after this batch.

File: client/client.go
Comment (KSDaemon) [reply client/client.go:850]: Yep. If I understand correctly, later after `msg, err := c.waitForReplyWithCancel(ctx, id, procedure, progChan)` if chan was opened — it is closed then

File: client/client.go
Comment (KSDaemon) [reply client/client.go]: Again, If I understand correctly, `select` in waitForReplyWithCancel upon ctx cancelation falls into `case <-ctx.Done()` so it is processed correctly and then returns here for the rest

File: client/client.go
Comment (KSDaemon) [reply client/client.go]: yes.  while receiving progressive results

File: client/client.go
Comment (KSDaemon) [reply client/client.go]: Hmm... But if I make it through the defer — I still need to check if its err'ed to return before the switch. So I don't see any profit. Or do you mean something else?

File: client/client.go
Comment (KSDaemon) [reply client/client.go]: O yeah! Forgot about that go idiomatic way :) Thnx!)

File: client/client.go
Comment (KSDaemon) [reply client/client.go]: Well, it may return an error, which is never checked for across the whole project. So I'll leave it untouched for now

**Comment by KSDaemon** [comment]
Well, at least it works :) 
But I'm sure, there are places that can be improved/refactored/made more optimal.

File: client/client.go
Comment (zyga) [reply client/client.go]: Just that you can have this code written once. I will re-read the diff with fresh eyes in the morning.

Comment (zyga) [reply client/client.go]: Just that you can have this code written once. I will re-read the diff with fresh eyes in the morning.

File: client/client.go
Comment (zyga) [reply client/client.go]: I mean, it's weird. Either:

1) We should change `Send` not to return an `error` value.
2) Or we should return the error value it returns.
3) Or we should use something like `_ = c.sess.Send(message)` to explicitly *ignore* the error value.

What we have now looks like an omission, not deliberate design.

Comment (KSDaemon) [reply client/client.go]: Yes, you are right of course!
Well, i'll think, maybe adding explicit ignore — is ok. As it is just sending message through the wire using underlying connection.

File: client/client.go
Comment (bboozzoo) [inline client/client.go]:
```suggestion
	err := c.prepareCallPayloadMessage(message, options, args, kwargs)
	if err != nil {
```

File: client/client.go
Comment (bboozzoo) [inline client/client.go:871]: looked at some other code, intersting that errors on Send() aren't checked, so if sending fails the caller of the client code actually has to wait for wait for reply to report an error

File: client/client.go
Comment (bboozzoo) [inline client/client.go]: I'm wondering why the push model (i.e. caller pushing more data to the progressive call through the returned callback), rather than a pull model (where the client would request more data from the caller by calling a callback provided in the input arguments).

I think the push model makes it a little awkward and forces the need for the final result callback, as the call side of `CallProgressive` and the result are separated if I'm not mistaken. So it is possible that CallProgressive would return a nil error and then you only get the error through the `finalrescb`.

```
resC := make(chan err, 1)
resultCb := func(*wamp.Result, error) {
    resC <- err
}
resultProgressCb := func(*wamp.Result) {
   // got progress
}
res, err := c.CallProgressive(ctx, "com.foo.bar", nil, nil, nil, resultCb, resultProgressCb)
if err != nil {
    // launching the call failed
}
// push push push
...
// wait for result
if err := <-resC; err != nil {
    // the call failed
}
```

If OTOH we used a pull model, the client code would could look like this:

```
sendMore := func() (options wamp.Dict, args wamp.List, kwargs wamp.Dict, isLast bool, err  error) {
    return nil, nil, nil, nil
}
resultProgressCb := func(*wamp.Result) {
   // got progress
}
res, err := c.CallProgressive(ctx, "com.foo.bar", nil, nil, nil, resultProgressCb, sendMore)
if err != nil {
    // the call failed
}
```

I think it's a bit simpler to wrap one's head around it.

File: router/dealer.go
Comment (bboozzoo) [inline router/dealer.go]:
```suggestion
	// otherwise we must reuse runtime-data. E.g. not to generate new ID,
	// so that any call progress is reaching the same caller we chose initially
```

**Review by bboozzoo — COMMENTED** [review]
Sorry for late review. Some comments inline.

File: client/client.go
Comment (bboozzoo) [reply client/client.go]: A slightly different approach to the pull model we discussed in the call. which also addresses the problem with the client code knowing when is the final call. Let's say we have the API like so:

```go
type SendProgressHandler func(ctx context.Context) (args wamp.List, kwargs wamp.Dict, options wamp.Dict, err error)

// CallProgressive makes progressive calls to the procedure corresponding to
// the given URI.
//
func (c *Client) CallProgressive(ctx context.Context, procedure string, prog SendProgressHandler, progcb ProgressHandler) (*wamp.Result, error) { 
	...
}
```

When the caller returns `options[wamp.OptProgress] = true` in the return arguments from `SendProgressHandler`, this means that it's a intermediate invocation. The lack of `"progress"=true` option means it's the final invocation. So in this scenario, the progressive call like this:

```go
progCb := func(ctx context.Context) (args wamp.List, kwargs wamp.Dict, options wamp.Dict, err error) {
	return wamp.List{"foo"}, nil, nil, nil
}
res, err := c.CallProgressive(ctx, "com.foo.bar", callProg, nil)
```

is equivalent to a simple call:

```go
res, err := c.Call(ctx, "com.foo.bar", wamp.List{"foo"}, nil, nil, nil)
```

while this would be a progressive call with 2 invocations, where 2nd invocation is final (and still carries data):

```go
cnt := 0
progCb := func(ctx context.Context) (args wamp.List, kwargs wamp.Dict, options wamp.Dict, err error) {
	if cnt == 0 { 
		cnt++
		return wamp.List{"the end"}, nil, wamp.Dict{wamp.OptProgress: true}, nil
	}
	return wamp.List{"the end"}, nil, nil, nil
}
res, err := c.CallProgressive(ctx, "com.foo.bar", callProg, nil)
```

File: router/dealer.go
Comment (KSDaemon) [reply router/dealer.go]: 👍🏻  :)

**Comment by KSDaemon** [comment]
@bboozzoo Refactored to pull model

**Comment by KSDaemon** [comment]
And we still have a question of `Send` error checking on the plate. There are tons of such places. Let's think of this. But I suggest working around that in different PR.

File: client/client.go
Comment (bboozzoo) [inline client/client.go]:
```suggestion
// SendProgressiveData is a callback function for feeding progressive call data.
// This function is called by the client to retrieve payload from business side.
// The returned options indicate whether the data chunk returned by the callback is final,
// in which case the wamp.OptProgress option is false or unset, or intermediate with more
// data to come, in which case wamp.OptProgress option must be set to true.
```

File: client/client.go
Comment (bboozzoo) [inline client/client.go]:
```suggestion
	case *wamp.Result:
		err := c.prepareCallResultMessage(msg)
```

File: client/client.go
Comment (bboozzoo) [inline client/client.go]:
```suggestion
// CallProgressive makes a progressive call to the procedure corresponding to
```

File: client/client.go
Comment (bboozzoo) [inline client/client.go]:
```suggestion
// See the documentation of SenProgressiveData for more details about feeding the data
// to a progressive call.
```

File: client/client.go
Comment (bboozzoo) [inline client/client.go]:
```suggestion
	callInProgress, _ := options[wamp.OptProgress].(bool)
```

File: client/client.go
Comment (bboozzoo) [inline client/client.go]:
```suggestion
				callInProgress, _ = options[wamp.OptProgress].(bool)
```

File: client/client_test.go
Comment (bboozzoo) [inline client/client_test.go]: is the sleep necessary?

File: client/client_test.go
Comment (bboozzoo) [inline client/client_test.go]: same thing

File: client/client_test.go
Comment (bboozzoo) [inline client/client_test.go]:
```suggestion
		mu.Lock()
		defer mu.Unlock()
		progressiveResults = append(progressiveResults, result.Arguments[0].(int))
```

File: client/client_test.go
Comment (bboozzoo) [inline client/client_test.go:843]: it would be good to verify that progressiveIncPayload has the same order as callArgs

File: client/client_test.go
Comment (bboozzoo) [inline client/client_test.go:934]: same thing about verifying the order of callArgs and progressiveResults to match

File: wamp/options.go
Comment (bboozzoo) [inline wamp/options.go:24]: unrelated?

Comment (KSDaemon) [reply wamp/options.go:24]: Related, but not yet supported :)

File: client/client_test.go
Comment (KSDaemon) [reply client/client_test.go]: well, yeah, seems no need. Removed it.
But then tests starts failing: payloads are not ordered. This means 
that something is not ordered... But i can't figure out what is it?

In `CallProgressive` the first one callback is fired and then `c.sess.Send` is called which in turn should just add a message to localpeer queue.
So all next `sendProg` calls that happening in the goroutine in loop should also just push a message to localpeer buffered channel....

File: wamp/options.go
Comment (KSDaemon) [reply wamp/options.go:24]: That is tricky and requires more work, so I postpone it for now.

Situation: shared registration of PRC with not sticked invocation policy (e.g. Round robin) + progressive call + longtime running callees.

For single call it is easy — callee is chosen once upon invocation and message is routed to it. Thats all.

But progressive call consists of multiple messages. 
So there can be 2 situations: 
* all chunks of data are routed to the same callee (treating the whole call as one). This is a default behaviour.
* Every new chunk triggers a callee selection based on the invocation policy by specifying `sticky=false`. In this case router should keep info about all callees involved into the call + in case of call cancelling — router should cancel all callees respecting cancelling option and so on... Not so easy :) 

So the second flow is not implemented for now.

File: transport/localpeer.go
Comment (bboozzoo) [inline transport/localpeer.go]:
```suggestion
	// message to the broker or dealer. Therefore, this channel can be
```

File: wamp/options.go
Comment (bboozzoo) [reply wamp/options.go:24]: Thanks for explaining.

File: client/client_test.go
Comment (bboozzoo) [reply client/client_test.go]: I'll try to play with the branch locally and see what's going on there.

File: client/client_test.go
Comment (bboozzoo) [reply client/client_test.go]: I guess this sleep can be dropped now too?

File: client/client.go
Comment (bboozzoo) [inline client/client.go]: I think this here should really just return a special internal error, eg. `errProtocolVolation` and Send followed by Perr.Close() should happen in the caller

File: client/client.go
Comment (bboozzoo) [inline client/client.go]:
```suggestion
	ctx := c.invHandlersCtxs[cliInvocation]
```

File: client/client.go
Comment (bboozzoo) [inline client/client.go:1599]: what does the timeout mean in this case? is it the timeout of a single invocation or the timeout for the whole request up until the final result?  if it's per invocation then there should be a new context for each invocation being sent to the handler

File: client/client.go
Comment (bboozzoo) [inline client/client.go:1642]:
```suggestion
						// a final or the only invocation
						c.sess.Lock()
```

File: client/client.go
Comment (bboozzoo) [inline client/client.go:1652]: I think this is the right place to create a WithTimeout context if the timeouts are per invocation

File: client/client.go
Comment (KSDaemon) [reply client/client.go:1599]: Nope, this can not be a timeout for a single chunk, because again — there is no correlation between sending data and receiving results. 
In progressive calls the invocation starts with first call and lasts until receiving the final result. All intermediate calls (from client side) are still bound to the same invocation.

So in general I don't see how making  different contexts for every progressive call chunks can make sense... 
1st thought: may be update the previous one if it is provided?... Then it can be useful. Kind of pushing the deadline :)

File: client/client.go
Comment (KSDaemon) [reply client/client.go:1642]: check my comment above ^^^

File: client/client.go
Comment (bboozzoo) [reply client/client.go:1599]: ok, so the timeout is for the whole complete invocation then? i.e. starts with the first invoke and must last until the final result?

Comment (KSDaemon) [reply client/client.go:1599]: Yep!

File: client/client.go
Comment (bboozzoo) [reply client/client.go]: Please check if you can split out the control path i.e. sending the error meesage to the peer and closing it from the actual code which only prepares the result before submitting this upstream.

**Review by bboozzoo — APPROVED** [review]
LGTM

**Comment by KSDaemon** [comment]
Good! Done!

**Pull request merged**
